src/Microsoft.IdentityModel.Logging/IIdentityLogger.cs

```diff
@@ -0,0 +1,49 @@
+﻿//------------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System.Diagnostics.Tracing;
+
+namespace Microsoft.IdentityModel.Logging
+{
+    /// <summary>
+    /// Interface for Logging.
+    /// </summary>
+    public interface IIdentityLogger
```

Comment:
There is no provision for turning PII logging on / off or for PII in general.

Reply:
Yes, we need some way of determining if Pii is enabled so the decision can be made if an expensive message containing Pii should be created or not.

Reply:
This is a good point about expensive messages, and a scenario that should be articulated. One way of doing it: https://gist.github.com/bgavrilMS/16d44dca6370588db0b95fc370e73557#file-iidentitylogger-cs-L32

Reply:
True. Within the IIdentityLogger currently there is no provision for PII because Wilson exposes a static IdentityEventSource.ShowPII boolean that developer can set to control if PII will be logged or not.

Reply:
[The console ](https://gist.github.com/bgavrilMS/16d44dca6370588db0b95fc370e73557#file-iidentitylogger-cs-L32)
=> This looks very promising for MSAL's requirements. I'll check with the team if we want to introduce a PII property that is associated with IIdentityLogger interface in Wilson.

Reply:
@sruke @trwalke above we seem to have identified two issues:
1. add logic to delay creation of messages until after determining the message will be logged (log level is in effect).
2. controlling PII. We currently have the switch on IdentityEventSource.ShowPII, if IIdentityLogger is independent of IdentityEventSource, then we will need logic.

Reply:
About (2)
Moving forward Wilson will introduce ILogScrubber whose implementation will contain a minimumDataClassificationLevel property. This property will decide the level of scrubbing and will decouple the concept of PII and Logging. Until then, IIdentityLogger will use IdentityEventSource.ShowPII to dictate if messages with PII can be logged. Spoke to @trwalke offline who will explore maintaining a PII property that is separate from IIdentityLogger. 

About (1)
[Here](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/586bb1a47777860b0a517f9399e2b216136ffb8c/src/Microsoft.IdentityModel.Logging/LogHelper.cs#:~:text=if%20(Logger.IsEnabled(EventLevel.Verbose))) is where logger checks if the given log level is enabled before proceeding to scrub the log arguments and formatting the message.

Reply:
@sruke - the example you give for (1) is does not show how the logger would handle an expensive to compute message, This is more like: 

```csharp
Log(Action<string> messageProducer)
{
   if (Logger.IsEnabled())
   {
          message = messageProducer.Invoke();
          logger.Log(message);
   }
}

// usage: 
Log(() => "this is an expensive message");
```

This is how ASP.NET Core's logger does it as well, while maintaiing the ILogger / IIdentityLogger interface simple

Reply:
So logger.Log(message) expects a pre-formatted message that is scrubbed off of PII. 

The way this handled in Wilson is -
`
if (Logger.IsEnabled(EventLevel.Verbose)) // if given log level is enabled proceed to compute message
{
     LogEntry entry = WriteEntry(EventLevel.Verbose, null, message, args) // message and args are passed separately
     Logger.Log(entry);
}
`
[WriteEntry](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/586bb1a47777860b0a517f9399e2b216136ffb8c/src/Microsoft.IdentityModel.Logging/LogHelper.cs#:~:text=private%20static%20LogEntry%20WriteEntry(EventLevel%20level%2C%20Exception%20innerException%2C%20string%20message%2C%20params%20object%5B%5D%20args)) is responsible for generating a LogEntry

`
private static LogEntry WriteEntry(EventLevel level, Exception innerException, string message, params object[] args)
{
      LogEntry entry = new LogEntry();
      entry.EventLevel = level;
      entry.Message = args != null ? FormatInvariant(message, args) : message;
      return entry;
`

LogEntry in turn invokes [FormatInvariant](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/586bb1a47777860b0a517f9399e2b216136ffb8c/src/Microsoft.IdentityModel.Logging/LogHelper.cs#:~:text=public%20static%20string%20FormatInvariant(string%20format%2C%20params%20object%5B%5D%20args)) which handles the processing of the expensive message based on the value of IdentityModelEventSource.ShowPII

Reply:
This works, just make sure that message and args are not evaluated beforehand.

```csharp
if (Logger.IsEnabled(EventLevel.Verbose)) // if given log level is enabled proceed to compute message
{
     message = "Here is an expensive to compute message with arg0 {0} and arg1 {1} ";
     arg0 = ExpensiveOperation0();
     arg1 = ExpensiveOperation1();

     LogEntry entry = WriteEntry(EventLevel.Verbose, null, message, new[] {arg0, arg1}) // message and args are passed separately
     Logger.Log(entry);
}
```